script: Replace usage of `IntersectionObserverRootMargin` with Stylo's`IntersectionObserverMargin`

labros7107 opened the pull request:
@stevennovaryo 
Created wrapper for Stylo's IntersectionObserverMargin and cleaned up repeated code.
Testing: Code compiles and `./mach test-unit tests/unit/style/` doesn't have any errors. intersectionobserver.rs is able to utilize the struct.
Fixes: https://github.com/servo/servo/issues/35907

---------------------------
Signed-off-by: samir <samir.khan720a@gmail.com>

labros7107 committed:
Created wrapper for Stylo's IntersectionObserverMargin and cleaned up
repeated code. Additionally moved resolve_percentages_with_basis() to
work with wrapper.
---------------------------
Signed-off-by: samir <samir.khan720a@gmail.com>

stevennovaryo reviewed (commented):
Looking good! Several suggestions to go.

labros7107 commented on components/script/dom/intersectionobserverrootmargin.rs:
```diff
@@ -2,97 +2,53 @@
-        writer.item(&rect.3)
+        self.0.to_css(dest)
     }
 }
```
Makes sense to me, keeping in mind I may not fully understand the bigger picture 😅.

Would this mean instead of using a wrapper at all we just change `intersectionobserver.rs` to directly use `style::values::specified::intersection_observer::IntersectionObserverMargin`? In which case do we remove the `intersectionobserverrootmargin.rs` file completely and move `resolve_percentages_with_basis` to `intersectionobserver.rs`

Doing this might resolve the other suggestions as well.

stevennovaryo replied:
Yes, exactly! Just keep in mind about the documentations and the naming. 👍

stevennovaryo commented on components/script/dom/intersectionobserverrootmargin.rs:
```diff
@@ -2,97 +2,53 @@
 }
 
-// TODO(stevennovaryo): move this to the wrapper later
 impl IntersectionObserverRootMargin {
```
Do you mean like this? https://github.com/servo/servo/pull/38519#discussion_r2261794933

labros7107 replied:
I have implemented the changes by referencing Stylo's implementation of `IntersectionObserverMargin` where applicable and moved `resolve_percentages_with_basis` under `IntersectionObserver`. Removed the `intersectionobserverrootmargin.rs` and changes `mod.rs` to not reference it.

labros7107 committed:
Stylo's InterectionObserverMargin struct is now visible, Servo can
call it directly instead of using a local clone of the struct. All
references to the local clone of the struct has now been moved to the
upstream implementation. Additionally, custom functions added to the
local clone has been reimplemented directly within
intersectionobserver.rs

Signed-off-by: samir <samir.khan720a@gmail.com>

labros7107 committed:
Removed intersectionobserverrootmargin.rs as it was a clone of Stylo's
IntersectionObserverMargin struct. As Stylo's implementation is now
visible, it is no longer required. Changed mod.rs file to not reference
deleted file.
Signed-off-by: samir <samir.khan720a@gmail.com>

labros7107 committed:
Merge branch 'servo:main' into issue-35907

labros7107 committed:
Ran mach fmt and test-tidy. No errors.

Signed-off-by: samir <samir.khan720a@gmail.com>
